eva-accession-clustering/src/test/java/uk/ac/ebi/eva/accession/clustering/batch/io/ClusteringMongoReaderRetryTest.java

```diff
@@ -0,0 +1,158 @@
+import org.springframework.batch.item.ExecutionContext;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.convert.MongoConverter;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import uk.ac.ebi.eva.accession.core.model.eva.SubmittedVariantEntity;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static uk.ac.ebi.eva.accession.clustering.batch.io.ClusteringMongoReader.MAX_RETRIES;
+
+@RunWith(SpringRunner.class)
+public class ClusteringMongoReaderRetryTest {
+
+    private static final String ASSEMBLY = "GCA_000000001.1";
+
+    private static final int CHUNK_SIZE = 5;
+
+    private ClusteringMongoReader nonClusteredVariantReader;
+
+    @MockBean(answer = Answers.RETURNS_DEEP_STUBS)
+    private MongoTemplate mongoTemplate;
+
+    @Before
+    public void setUp(){
+        nonClusteredVariantReader = spy(new ClusteringMongoReader(mongoTemplate, ASSEMBLY, CHUNK_SIZE, false));
+    }
+
+    @After
+    public void tearDown() {
+        nonClusteredVariantReader.close();
+    }
+
+    @Test
+    public void readSucceedsWhenCursorExceptionThrownOnce() {
+        MongoCursor mockCursor = mock(MongoCursor.class);
+        when(mockCursor.hasNext()).thenReturn(true, true, true, true, false);
+        Document doc1 = new Document("id_", "1");
+        Document doc2 = new Document("id_", "2");
+        Document doc3 = new Document("id_", "3");
+        when(mockCursor.next()).thenReturn(doc1)
+                               .thenThrow(MongoCursorNotFoundException.class)
+                               .thenReturn(doc2, doc3);
+        openReaderWithMockCursor(mockCursor);
+
+        List<SubmittedVariantEntity> variants = readIntoList(nonClusteredVariantReader);
+        assertEquals(variants.stream().map(sve -> sve.getId()).collect(Collectors.toList()),
+                     Arrays.asList("1", "2", "3"));
+        // next() called once per document, plus 1 with a failure
+        verify(mockCursor, times(4)).next();
+        // initializeReader() called once at the start and once during the retry
+        verify(nonClusteredVariantReader, times(2)).initializeReader();
+    }
+
+    @Test
+    public void readFailsWhenOtherExceptionThrownOnce() {
+        MongoCursor mockCursor = mock(MongoCursor.class);
+        when(mockCursor.hasNext()).thenReturn(true, true, true, true, false);
+        Document doc1 = new Document("id_", "1");
```

Comment:
Are these "id_" as opposed to "_id" for a particular reason? The reader sort does seem to work with the correct attribute "_id"..

Reply:
no reason except that I screwed it up... thanks!